### ISSUE TITLE
Add update conflict algorithm to insert helper (UPSERT)

### ISSUE THREAD

**Issue by bscottsmith** [tekartik/sqflite#436]
Enhancing the insert helper method to access "UPSERT" functionality would be extremely useful.
I can not use the insert helper with "ConflictAlgorithm.replace" because that will delete the conflicting row, and I have several foreign tables with CASCADE DELETE. I need the primary key to remain the same, and all foreign references kept intact.
Adding a "ConflictAlgorithm.update" to the insert helper would be ideal, giving access to the "ON CONFLICT DO UPDATE..." clause of the SQLite INSERT statement.
```
  Future<void> _dbInsert(Account account) {
    return db.insert(_dbTable, account.toJson(),
        conflictAlgorithm: ConflictAlgorithm.update);
  }
```

**Comment by bscottsmith** [comment]
See also #59, #102, and #217.

**Comment by alextekartik** [comment]
As far as I know this does not exist in SQLite: https://sqlite.org/lang_conflict.html

Unfortunately you have create a transaction, check if a record exist to either update it or insert a new one.

**Comment by bscottsmith** [comment]
It does, as an "ON CONFLICT" syntax addition to INSERT:
[https://www.sqlite.org/lang_UPSERT.html](url)

**Comment by alextekartik** [comment]
I see; indeed that could be useful...Don't know how the API could look like to:
- support the conflict column,
- support where clause in the update statement
without having an ugly API for a very specific usage.

Helper are for simple queries since it is hard to make a full featured sql builder, maybe some 3rd party package can help. In the meantime best is to use a rawInsert statement.

**Comment by bscottsmith** [comment]
Agreed, the where clause would make it messy, but that would rarely be used (according to the docs). Perhaps rather than extending the existing insert helper, it would be better to create a new upsert helper:
```
Future<void> _upsertAccount(Account account) {
  return db.upsert(_dbTable, account.toJson(),
      conflictColumn: "id",
      conflictAlgorithm: UpsertConflictAlgorithm.update);
}
```
So, given a table `account`, with columns `id`, `name`, `phone`, and `address`, the raw SQL statement generated from the above helper would be:
```
INSERT INTO account (id, name, phone, address) 
  VALUES (1234, 'John Doe', '+15125551212', '123 Main St.')
  ON CONFLICT (id) DO UPDATE 
  SET name=excluded.name, phone=excluded.phone, address=excluded.address
```
Notice that I did not include the `conflictColumn` in the UPDATE clause, although this is not necessary.
Or, if `conflictAlgorithm: UpsertConflictAlgorithm.nothing` is used, the SQL would be:
```
INSERT INTO account (id, name, phone, address) 
  VALUES (1234, 'John Doe', '+15125551212', '123 Main St.')
  ON CONFLICT (id) DO NOTHING
```
Although all this will be moot unless you are embedding version 3.24.0 (from 2018) or better to support the feature...

**Comment by omidraha** [comment]
Related [link](https://stackoverflow.com/questions/15277373/sqlite-upsert-update-or-insert).

**Comment by cassioseffrin** [comment]
Agreed with @bscottsmith UPSERT (new helper) will be very very appreciated.

Do you have some plans to support it?

**Comment by fldvlpr** [comment]
@alextekartik @davidmartos96 @joshpetit @marcoms @mit-mit @sethladd @kevmoo @bscottsmith Does this package support on conflict sqlite raw query?

I was trying following query in my sqlite DB Browser and it works fine.

`INSERT Into tbl_users(_id, category, description, amount) values(6, "kkk", "mm", 140.00) ON Conflict(_id) DO UPDATE SET category = "kkk", description = "mmm", amount = 140.00
` 

But, when trying to run query using sqflite package,

`final result = await db.rawQuery("INSERT Into tbl_users(_id, category, description, amount) values(1, \"kkk\", \"mm\", 140.00) ON Conflict(_id) DO UPDATE SET category = \"kkk\", description = \"mmm\", amount = 140.00");`

It gives me an error like, 

```
near "ON": syntax error
E/flutter (21228): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: DatabaseException(near "ON": syntax error (code 1): , while compiling: INSERT Into tbl_users(_id, category, description, amount) values(1, "kkk", "mm", 140.00) ON Conflict(_id) DO UPDATE SET category = "kkk", description = "mmm", amount = 140.00) sql 'INSERT Into tbl_users(_id, category, description, amount) values(1, "kkk", "mm", 140.00) ON Conflict(_id) DO UPDATE SET category = "kkk", description = "mmm", amount = 140.00' args []}
```

Let me know if it's supported in this package or not.

Thanks.

**Comment by alextekartik** [comment]
No need to ping the whole dart team!

sqflite on iOS and Android uses whatever SQLite bundled on the platform. It seems the UPSERT (ON CONFLICT) feature was added in SQLite 3.24 which might not be present in some Android version: See https://stackoverflow.com/a/4377116/563979

You can do a "SELECT sqlite_version()" to check the version.

I see 2 solutions:
- Do not use on CONFLICT and instead use transaction to catch the error on the insert and then perform the update
- Try sqflite_ffi along with sqlite3 and sqlite3 flutter libs. I have not tried it much but it should work

**Comment by tomekit** [comment]
I am wondering how does the Android can support this `insertWithOnConflict` despite old SQLite versions, as this is available already in API 8:
https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase#insertWithOnConflict(java.lang.String,%20java.lang.String,%20android.content.ContentValues,%20int)

and supported conflict modes:
https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase#CONFLICT_ABORT

**Comment by JediPixels** [comment]
Check on the status of this UPSERT feature, has it been implemented?

```
await db.insert('notes', notes.toJson(), conflictAlgorithm: ConflictAlgorithm.replace);
```

Thank you.

**Comment by alextekartik** [comment]
You might still have to check the SQLite version before using upsert when using sqflite but you can use it when using sqflite_common_ffi
